From a54dbc5294352cbf7b13efd62034fe6fff485502 Mon Sep 17 00:00:00 2001
From: John Chung <john.chung@mic.com.tw>
Date: Fri, 19 Feb 2021 14:46:33 +0800
Subject: [PATCH] Support FRU configuration in JSON format

Support FRU device id attribute
Support FRU device size attribute
Support FRU device name attribute
Support FRU device read-only attribute
---
 include/storagecommands.hpp |   7 ++
 src/storagecommands.cpp     | 182 ++++++++++++++++++++++++++++++++++++++------
 2 files changed, 166 insertions(+), 23 deletions(-)

diff --git a/include/storagecommands.hpp b/include/storagecommands.hpp
index 3d28f48..59c29b5 100755
--- a/include/storagecommands.hpp
+++ b/include/storagecommands.hpp
@@ -55,6 +55,13 @@ struct GetSDRReq
 };
 #pragma pack(pop)
 
+struct deviceInfo
+{
+    uint8_t bus;
+    uint8_t address;
+    bool readOnly;
+};
+
 enum class SdrRepositoryInfoOps : uint8_t
 {
     allocCommandSupported = 0x1,
diff --git a/src/storagecommands.cpp b/src/storagecommands.cpp
index d1a162e..ff95a99 100755
--- a/src/storagecommands.cpp
+++ b/src/storagecommands.cpp
@@ -131,12 +131,21 @@ constexpr static const uint8_t deassertionEvent = 0x80;
 static std::vector<uint8_t> fruCache;
 static uint8_t cacheBus = 0xFF;
 static uint8_t cacheAddr = 0XFF;
+static uint32_t cacheSize = 256;
 
 std::unique_ptr<phosphor::Timer> cacheTimer = nullptr;
 
 // we unfortunately have to build a map of hashes in case there is a
 // collision to verify our dev-id
-boost::container::flat_map<uint8_t, std::pair<uint8_t, uint8_t>> deviceHashes;
+boost::container::flat_map<uint8_t, std::tuple<uint8_t, uint8_t, std::string>> deviceHashes;
+boost::container::flat_map<uint8_t, struct deviceInfo> deviceMap;
+
+// JSON of FRU configuration
+constexpr const char* fruConfig =
+    "/usr/share/entity-manager/configurations/fru.json";
+
+// FRU inventory size
+boost::container::flat_map<uint8_t, uint32_t> deviceSizes;
 
 void registerStorageFunctions() __attribute__((constructor));
 
@@ -215,8 +224,12 @@ ipmi_ret_t replaceCacheFru(uint8_t devId)
     // if timer is running, stop it and writeFru manually
     else if (timerRunning && cacheTimer->isRunning())
     {
+        bool timerIsRunning = cacheTimer->isRunning();
         cacheTimer->stop();
-        writeFru();
+        if(timerIsRunning)
+        {
+            writeFru();
+        }
     }
 
     std::shared_ptr<sdbusplus::asio::connection> dbus = getSdBus();
@@ -237,6 +250,8 @@ ipmi_ret_t replaceCacheFru(uint8_t devId)
     }
 
     deviceHashes.clear();
+    deviceSizes.clear();
+    deviceMap.clear();
 
     // hash the object paths to create unique device id's. increment on
     // collision
@@ -251,35 +266,125 @@ ipmi_ret_t replaceCacheFru(uint8_t devId)
 
         auto busFind = fruIface->second.find("BUS");
         auto addrFind = fruIface->second.find("ADDRESS");
+        auto indexFind = fruIface->second.find("INDEX");
         if (busFind == fruIface->second.end() ||
-            addrFind == fruIface->second.end())
+            addrFind == fruIface->second.end() ||
+            indexFind == fruIface->second.end())
         {
             phosphor::logging::log<phosphor::logging::level::INFO>(
-                "fru device missing Bus or Address",
+                "fru device missing Bus or Address or Index.",
                 phosphor::logging::entry("FRU=%s", fru.first.str.c_str()));
             continue;
         }
 
         uint8_t fruBus = std::get<uint32_t>(busFind->second);
         uint8_t fruAddr = std::get<uint32_t>(addrFind->second);
+        uint8_t fruIndex = std::get<uint32_t>(indexFind->second);
+
+        bool readOnly = false;
+        auto readOnlyFind = fruIface->second.find("READONLY");
+        if (readOnlyFind != fruIface->second.end())
+        {
+            readOnly = std::get<bool>(readOnlyFind->second);
+        }
+
+        struct deviceInfo newDevice = {fruBus, fruAddr, readOnly};
+
+        auto deviceFind = deviceMap.find(fruIndex);
+        if (deviceFind == deviceMap.end())
+        {
+            deviceMap.emplace(fruIndex, newDevice);
+        }
+        else
+        {
+            std::cerr << "Fru devices have the same index. Index: "
+                      << static_cast<uint32_t>(fruIndex)
+                      << ", Bus: " << static_cast<uint32_t>(fruBus)
+                      << ", Address: " << static_cast<uint32_t>(fruAddr)
+                      << "\n";
+        }
 
         uint8_t fruHash = 0;
+
+        // Set the fru default size to 256 bytes.
+        uint32_t fruSize = 256;
+        std::string deviceID = "";
+
         if (fruBus != 0 || fruAddr != 0)
         {
             fruHash = hasher(fru.first.str);
+
+            std::ifstream FruFile(fruConfig);
+
+            if(!FruFile)
+            {
+                std::cerr << "Failed to open FRU config file: " << fruConfig <<"\n";
+            }
+            else
+            {
+                auto data = nlohmann::json::parse(FruFile, nullptr, false);
+                if (data.is_discarded())
+                {
+                    std::cerr << "syntax error in " << fruConfig << "\n";
+                }
+                else
+                {
+                    int index = 0;
+
+                    while (!data["Fru"][index].is_null())
+                    {
+                        if (!data["Fru"][index]["Bus"].is_null() &&
+                            !data["Fru"][index]["Address"].is_null())
+                        {
+                            int bus = data["Fru"][index]["Bus"];
+                            std::string addrStr = data["Fru"][index]["Address"];
+                            int address = std::stoul(addrStr, nullptr, 16);
+                            // int address = atoi(addrStr.substr(addrStr.find("x", 0)+1).c_str());
+
+                            if (bus == fruBus && address == fruAddr)
+                            {
+                                if (!data["Fru"][index]["Size"].is_null())
+                                {
+                                    fruSize = static_cast<uint32_t>(data["Fru"][index]["Size"]);
+                                }
+
+                                if (!data["Fru"][index]["Index"].is_null())
+                                {
+                                    fruHash = static_cast<uint8_t>(data["Fru"][index]["Index"]);
+                                }
+
+                                if(!data["Fru"][index]["Name"].is_null())
+                                {
+                                    deviceID = static_cast<std::string>(data["Fru"][index]["Name"]);
+                                }
+
+                                break;
+                            }
+
+                        }
+                        index++;
+                    }
+                }
+
+                FruFile.close();
+            }
+
             // can't be 0xFF based on spec, and 0 is reserved for baseboard
-            if (fruHash == 0 || fruHash == 0xFF)
+            if (fruHash == 0xFF)
             {
                 fruHash = 1;
             }
         }
-        std::pair<uint8_t, uint8_t> newDev(fruBus, fruAddr);
+        std::tuple<uint8_t, uint8_t, std::string> newDev(fruBus, fruAddr, deviceID);
 
         bool emplacePassed = false;
         while (!emplacePassed)
         {
             auto resp = deviceHashes.emplace(fruHash, newDev);
             emplacePassed = resp.second;
+
+            deviceSizes.emplace(fruHash, fruSize);
+
             if (!emplacePassed)
             {
                 fruHash++;
@@ -299,16 +404,29 @@ ipmi_ret_t replaceCacheFru(uint8_t devId)
     }
 
     fruCache.clear();
+
+    auto deviceSizeFind = deviceSizes.find(devId);
+    if (deviceSizeFind == deviceSizes.end())
+    {
+        return IPMI_CC_SENSOR_INVALID;
+    }
+    cacheSize = deviceSizeFind->second;
+
     sdbusplus::message::message getRawFru = dbus->new_method_call(
         fruDeviceServiceName, "/xyz/openbmc_project/FruDevice",
         "xyz.openbmc_project.FruDeviceManager", "GetRawFru");
-    cacheBus = deviceFind->second.first;
-    cacheAddr = deviceFind->second.second;
+    cacheBus = std::get<0>(deviceFind->second);
+    cacheAddr = std::get<1>(deviceFind->second);
     getRawFru.append(cacheBus, cacheAddr);
     try
     {
         sdbusplus::message::message getRawResp = dbus->call(getRawFru);
         getRawResp.read(fruCache);
+
+        if(fruCache.size() < cacheSize)
+        {
+            fruCache.resize(cacheSize, 0xFF);
+        }
     }
     catch (sdbusplus::exception_t&)
     {
@@ -401,6 +519,21 @@ ipmi_ret_t ipmiStorageWriteFRUData(ipmi_netfn_t netfn, ipmi_cmd_t cmd,
     {
         return status;
     }
+
+    auto deviceFind = deviceMap.find(req->fruDeviceID);
+    if (deviceFind == deviceMap.end())
+    {
+        return IPMI_CC_SENSOR_INVALID;
+    }
+    auto& device = deviceFind->second;
+
+    if (device.readOnly == true)
+    {
+        std::cerr << "Fru " << static_cast<int>(req->fruDeviceID)
+                  << " is read only device\n";
+        return IPMI_CC_INVALID_FIELD_REQUEST;
+    }
+
     int lastWriteAddr = req->fruInventoryOffset + writeLen;
     if (fruCache.size() < lastWriteAddr)
     {
@@ -531,8 +664,8 @@ ipmi_ret_t getFruSdrs(size_t index, get_sdr::SensorDataFruRecord& resp)
         return IPMI_CC_INVALID_FIELD_REQUEST;
     }
     auto device = deviceHashes.begin() + index;
-    uint8_t& bus = device->second.first;
-    uint8_t& address = device->second.second;
+    uint8_t& bus = std::get<0>(device->second);
+    uint8_t& address = std::get<1>(device->second);
 
     ManagedObjectType frus;
 
@@ -582,20 +715,23 @@ ipmi_ret_t getFruSdrs(size_t index, get_sdr::SensorDataFruRecord& resp)
     {
         return IPMI_CC_RESPONSE_ERROR;
     }
-    std::string name;
-    auto findProductName = fruData->find("BOARD_PRODUCT_NAME");
-    auto findBoardName = fruData->find("PRODUCT_PRODUCT_NAME");
-    if (findProductName != fruData->end())
-    {
-        name = std::get<std::string>(findProductName->second);
-    }
-    else if (findBoardName != fruData->end())
-    {
-        name = std::get<std::string>(findBoardName->second);
-    }
-    else
+    std::string name = std::get<2>(device->second);
+    if(name.empty())
     {
-        name = "UNKNOWN";
+        auto findProductName = fruData->find("BOARD_PRODUCT_NAME");
+        auto findBoardName = fruData->find("PRODUCT_PRODUCT_NAME");
+        if (findProductName != fruData->end())
+        {
+            name = std::get<std::string>(findProductName->second);
+        }
+        else if (findBoardName != fruData->end())
+        {
+            name = std::get<std::string>(findBoardName->second);
+        }
+        else
+        {
+            name = "UNKNOWN";
+        }
     }
     if (name.size() > maxFruSdrNameSize)
     {
-- 
2.7.4

